### PR TITLE
Add context manager and layer wrapper for marking NVTX ranges

### PR DESCRIPTION
**Pull request by danieldk**
This PR adds the following:

* the `with_nvtx_range` layer that wraps any layer and marks the `forward` and `backprop` passes as an NVTX range. This can be helpful when profiling GPU performance of one or more layers.
* the `use_nvtx_range` context manager, which can be used to mark a scope as an NVTX range.

The ranges are used to mark executions in nvprof or NSight profiles and make it easier to see to which layers/code CUDA operations are associated.

### CODE REVIEW

File: thinc/layers/with_nvtx_range.py [hunk 1]
```diff
@@ -34,5 +32,6 @@ def backprop(dY: Any) -> Any:
     def init(_model: Model, X: Any, Y: Any) -> Model:
         return layer.initialize(X, Y)
 
-    return Model(f"nvtx_range({name})", forward, init=init, layers=[layer])
-
+    return Model(
+        f"nvtx_range({name})", forward, init=init, layers=[layer], shims=layer.shims
```
Comment (danieldk) [inline thinc/layers/with_nvtx_range.py:36]: Not sure if we want `shims=layers.shims` or `shims=[layers.shims[0]]` here?

File: thinc/layers/with_nvtx_range.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+
+def with_nvtx_range(
+    layer: Model,
+    name: Optional[str] = None,
+    *,
+    forward_color: int = -1,
+    backprop_color: int = -1,
+):
+    """Layer that adds wraps any layer and marks the forward and backprop
```
Comment (svlandeg) [inline thinc/layers/with_nvtx_range.py:14]:
```suggestion
    """Layer that wraps any layer and marks the forward and backprop
```

File: thinc/layers/with_nvtx_range.py (hunk 1 above)
Comment (svlandeg) [reply thinc/layers/with_nvtx_range.py:36]: I think `layer.shims` is fine?

File: thinc/util.py [hunk 3]
```diff
@@ -462,6 +462,16 @@ def data_validation(validation):
         yield
         DATA_VALIDATION.set(prev)
 
+@contextlib.contextmanager
+def use_nvtx_range(message: int, id_color: int = -1):
+    """Context manager to register the executed code as an NVTX range. The
+    ranges are can be used as markers in CUDA profiling."""
```
Comment (svlandeg) [inline thinc/util.py:468]:
```suggestion
    ranges can be used as markers in CUDA profiling."""
```

File: website/docs/api-layers.md [hunk 4]
```diff
@@ -1222,6 +1222,40 @@ model.initialize()
 https://github.com/explosion/thinc/blob/master/thinc/layers/with_debug.py
 ```
 
+### with_nvtx_range {#with_debug tag="function"}
```
Comment (svlandeg) [inline website/docs/api-layers.md:1225]:
```suggestion
### with_nvtx_range {#with_nvtx_range tag="function"}
```

File: website/docs/api-layers.md [hunk 5]
```diff
@@ -1222,6 +1222,40 @@ model.initialize()
...
+from thinc.api import Linear, with_nvtx_range
+
+model = with_nvtx_range(Linear(2, 5))
+model.initialize()
+```
+
+| Argument         | Type                   | Description                                                                                               |
+| ---------------- | ---------------------- | ----------------------------------------------------------------------------------------------------------|
+| `layer`          | <tt>Model</tt>         | The layer to wrap.                                                                                        |
+| `name`           | <tt>Optional[str]</tt> | Optional name for the wrapped layer, will be prefixed by `debug:`. Defaults to name of the wrapped layer. |
```
Comment (svlandeg) [inline website/docs/api-layers.md:1248]:
```suggestion
| `name`           | <tt>Optional[str]</tt> | Optional name for the wrapped layer. Defaults to the name of the wrapped layer. |
```
(actually the `debug:` bit isn't true for the debug layer either, as it does `debug(name)`)

**Review by svlandeg — COMMENTED** [review]
Looks good to me! I only marked some very minor typo stuff.

File: thinc/layers/with_nvtx_range.py (hunk 1 above)
Comment (svlandeg) [reply thinc/layers/with_nvtx_range.py:36]: I think `layer.shims` is fine?

**Comment by danieldk** [comment]
> Looks good to me! I only marked some very minor typo stuff.

Thanks! All fixed.

**Pull request merged**